### PR TITLE
Feature/Restructuring OnResetTimeout

### PR DESCRIPTION
**Pull request by ghost**
Implements [#2422](https://github.com/smartdevicelink/sdl_requirements/issues/112) 

This PR is **ready** for review.

### Summary
Implementation of ResetTimeout

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: app/model/sdl/Abstract/Model.js [hunk 1]
```diff
@@ -1068,7 +1065,42 @@ SDL.SDLModel = Em.Object.extend({
...
+  vrTimeout: function(timer) {
+    self = SDL.SDLModel;
+    clearTimeout(self.timeout);
+    self.timeout = setTimeout(function() {
+        if (SDL.SDLModel.data.VRActive) {
+          if (SDL.SDLModel.data.vrActiveRequests.vrPerformInteraction) {
+            SDL.SDLController.vrInteractionResponse(
+              SDL.SDLModel.data.resultCode['TIMED_OUT']
+            );
+            // SDL.ResetTimeoutPopUp.resetTimeoutRPCs.removeObject('VR.PerformInteraction');
```
Comment (BSolonenko) [inline app/model/sdl/Abstract/Model.js]: @ValeriiMalkov please remove commented code.

File: app/view/sdl/shared/interactionChoicesView.js [hunk 2]
```diff
@@ -149,14 +150,18 @@ SDL.InteractionChoicesView = SDL.SDLAbstractView.create(
...
       if (this.timeout) {
         clearTimeout(this.timer);
         var self = this;
-        this.timer = setTimeout(
+        self.timer = setTimeout(
           function() {
-            self.deactivate('TIMED_OUT');
-          }, this.timeout
+            SDL.InteractionChoicesView.deactivate('TIMED_OUT');
+            // SDL.SDLAbstractView.deactivate();
```
Comment (BSolonenko) [inline app/view/sdl/shared/interactionChoicesView.js]: @ValeriiMalkov please remove commented code.

File: css/general.css [hunk 3]
```diff
@@ -166,6 +166,78 @@ div {
...
+}
+
+#ResetTimeoutPopUp .contextView .textArea {
+    width: 280px;
+    height: 34px;
+    color: white;
+    margin: 10px;
+    background-color: black;
+}
+#ResetTimeoutPopUp .contextView .list {
```
Comment (BSolonenko) [inline css/general.css]: @ValeriiMalkov please add white space.

File: css/general.css [hunk 4]
```diff
@@ -166,6 +166,78 @@ div {
...
+    height: 140px;
+    float: left;
+    position: relative;
+    text-shadow: 1px 1px 2px black, 0 0 6px rgba(255, 0, 0, 0.51), 0 13px 10px black;
+ }
+ #ResetTimeoutPopUp .contextView .list .component {
```
Comment (BSolonenko) [inline css/general.css]: @ValeriiMalkov please add white space.

File: app/model/sdl/Abstract/Model.js [hunk 5]
```diff
@@ -1020,15 +1025,23 @@ SDL.SDLModel = Em.Object.extend({
...
 
     var appID = message.params.appID;
-
+    var defaultTimeout = 10;
+    SDL.ResetTimeoutPopUp.timeoutSeconds = defaultTimeout;
     setTimeout(function() {
         if (SDL.SDLModel.data.vrActiveRequests.vrPerformInteraction) { // If VR PerformInteraction session is still active
           SDL.SDLModel.onPrompt(message.params.timeoutPrompt);
+          SDL.ResetTimeoutPopUp.timeoutSeconds['VR.PerformInteraction'] =
+           (message.params.timeout -(message.params.timeout - 2000))/1000;
```
Comment (AKalinich-Luxoft) [inline app/model/sdl/Abstract/Model.js]: @ValeriiMalkov why do you need -2000? Please move these magic numbers to a named constants

File: app/view/sdl/AlertPopUp.js [hunk 6]
```diff
@@ -227,17 +229,35 @@ SDL.AlertPopUp = Em.ContainerView.create(
...
-                                                                        // for
-                                                                        // Alert
-                                                                        // popUp
-      clearTimeout(this.timer);
-      this.timer = setTimeout(
+      this.setTimerUI(message.duration ? message.duration : this.defaultTimeout);
+      
+    },
+
+    setTimerTTS: function(time){
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/AlertPopUp.js]: @ValeriiMalkov add description

File: app/view/sdl/AlertPopUp.js [hunk 7]
```diff
@@ -227,17 +229,35 @@ SDL.AlertPopUp = Em.ContainerView.create(
...
+      var self = SDL.AlertPopUp;
+      self.set('ttsTimeout', time);
+      clearTimeout(self.ttsTimer);
+      self.ttsTimer = setTimeout(
+        function() {
+          clearTimeout(self.ttsTimer);
+        }, self.ttsTimeout
+      );
+    },
+    /*
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/AlertPopUp.js]: @ValeriiMalkov add space between functions

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+            'listOfRPC'
+        ],
+        content: "",
+        listOfRPC: SDL.List.extend({
+
+            elementId: 'select_items',
+            disableScrollbar: true,
+            value: true,
+        
+            /** Items */
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov remove this comment

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+        action: 'resetTimeout',
+        target: 'parentView',
+    }),
+    timerLabel: SDL.Label.extend({
+        elementId: 'timerLabel',
+        classNames: 'timerLabel',
+        contentBinding: 'parentView.timeoutString'
+    }),
+
+    player: SDL.AudioPlayer.create(),
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov why do you need player in this prompt?

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+     * ActivatePopUp function. activates pop-up
+     */
+    ActivatePopUp: function(){
+        length = this.resetTimeoutRPCs.length;
+        if(1 < length) {
+            this.addCheckBox(length);
+        }
+        this.set('isVisible', true);
+        this.set('active', true);
+        if('{}' == JSON.stringify(this.timeoutSeconds)) {
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov what is this?

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+            this.setDefaultTimeout();
+            this.resetTimeOutLabel();
+        }
+
+        const self = this;
+        this.timer = setInterval(
+            function() {
+                if(1 < length){
+                    timeout = '';
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov rename this variable to `message` and add `var` specifier

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+        this.set('callbacks', {});
+        this.set('resetTimeoutCallback', {});
+        this.set('resetTimeoutRPCs', []);
+        this.contextView.set('content',"")
+        this.player.stopPlaying();
+        this.player.clearFiles();
+    },
+
+    /**
+     * resetTimeoutMoreThenOne function for reset timeout if RPC more then one
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov than

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+     */
+    resetTimeoutMoreThenOne: function() {
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov `resetMoreThanOneTimeout`

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+        if(1 < length) {
+            this.resetTimeoutMoreThenOne();
+            return;
+        }
+        method = this.resetTimeoutRPCs[0];
+        element = document.getElementById(method + 'checkBox');
+        if(null !== element){
+            if(element.checked) {
+                reset();
+                return;
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov please rework this condition to get something like:
```
if (...) {
reset()
}
```

File: app/view/sdl/ResetTimeoutPopUp.js [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+            if(0 == self.resetTimeoutRPCs.length) {
+                self.DeactivatePopUp();
+            }
+            return;
+        }
+
+        var method = self.resetTimeoutRPCs[0];
+        element = document.getElementById(method + 'checkBoxSend');
+        
+        if(null !== element) {
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/ResetTimeoutPopUp.js]: @ValeriiMalkov please rework this condition for something like:
```
if (...) {
DeactivatePopUp(method);
}
```

File: app/view/sdl/shared/sliderView.js [hunk 16]
```diff
@@ -82,34 +82,39 @@ SDL.SliderView = SDL.SDLAbstractView.create(
           SDL.SDLModel.data.resultCode.SUCCESS, this.get('sliderRequestId'),
           this.get('adjustControl.sliderValue.value')
         );
-      } else {
+      } else if(timeout === null) {}
```
Comment (AKalinich-Luxoft) [inline app/view/sdl/shared/sliderView.js]: @ValeriiMalkov use negative condition like `else if (timeout !== null) {` and remove redundant else if

File: ffw/UIRPC.js [hunk 17]
```diff
@@ -159,7 +160,19 @@ FFW.UI = FFW.RPCObserver.create(
...
+                SDL.ResetTimeoutPopUp.extendResetTimeoutRPCs([request.method]);
+                SDL.ResetTimeoutPopUp.extendResetTimeoutCallBack(SDL.AlertPopUp.setTimerUI , request.method);
+                SDL.ResetTimeoutPopUp.expandCallbacks(function(){
+                  SDL.AlertPopUp.deactivate('timeout');
+                }, request.method);
+                if(request.params.duration)
+                  SDL.ResetTimeoutPopUp.set('timeoutSeconds',
+                    {'UI.Alert': request.params.duration/1000,
+                      'TTS.Speak': request.params.duration/1000});
+
```
Comment (AKalinich-Luxoft) [inline ffw/UIRPC.js]: @ValeriiMalkov remove extra space

File: app/model/sdl/Abstract/Model.js (hunk 1 above)
Comment (ValeriiMalkov) [reply app/model/sdl/Abstract/Model.js]: @BSolonenko Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/shared/interactionChoicesView.js (hunk 2 above)
Comment (ValeriiMalkov) [reply app/view/sdl/shared/interactionChoicesView.js]: @BSolonenko Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: css/general.css (hunk 4 above)
Comment (ValeriiMalkov) [reply css/general.css]: @BSolonenko Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/model/sdl/Abstract/Model.js (hunk 5 above)
Comment (ValeriiMalkov) [reply app/model/sdl/Abstract/Model.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/AlertPopUp.js (hunk 6 above)
Comment (ValeriiMalkov) [reply app/view/sdl/AlertPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/AlertPopUp.js (hunk 7 above)
Comment (ValeriiMalkov) [reply app/view/sdl/AlertPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 8 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 9 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 13 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 14 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 10 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft It is for checking that the container is empty

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 11 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 12 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 15 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/shared/sliderView.js (hunk 16 above)
Comment (ValeriiMalkov) [reply app/view/sdl/shared/sliderView.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: ffw/UIRPC.js (hunk 17 above)
Comment (ValeriiMalkov) [reply ffw/UIRPC.js]: @AKalinich-Luxoft  Fixed in commit a8caa0a45c6de14a6fce5bba659fdbabbde989c9

File: app/view/sdl/ResetTimeoutPopUp.js (hunk 10 above)
Comment (ValeriiMalkov) [reply app/view/sdl/ResetTimeoutPopUp.js]: @AKalinich-Luxoft That was fixed and forced update

**Pull request closed without merging**